### PR TITLE
gh-117045: Add code object to function version cache

### LINKED ISSUE #109998: Inconsistency in whether function setters call `_PyFunction_SetVersion`
# Bug report

The function object has a number of settable attributes. For the following, setting the value calls `_PyFunction_SetVersion(op, 0)`:

- `__defaults__`
- `__kwdefaults__`
- `__code__`
- `__annotations__`

But for the following, there is no such call:

- `__dict__`
- `__name__`
- `__qualname__`
- `__type_params__`

I don't think any of these four should affect the interpreter loop, but neither should `__annotations__`. Should these setters also set the version for consistency, or should `__annotations__` not set it as changing the annotations shouldn't affect the runtime?

cc @markshannon who added this in #27078.

### PR DESCRIPTION
**Pull request by gvanrossum**
## What?

Changes to the function version cache:

- In addition to the function object, also store the code object, and allow the latter to be retrieved even if the function has been evicted.
- Stop assigning new function versions after a critical attribute (e.g. `__code__`) has been modified; the version is permanently reset to zero in this case.
- Changes to `__annotations__` are no longer considered critical. (This fixes gh-109998.)

Changes to the Tier 2 optimization machinery:

- If we cannot map a function version to a function, but it is still mapped to a code object, we continue projecting the trace. The operand of the `_PUSH_FRAME` and `_POP_FRAME` opcodes can be either NULL, a function object, or a code object with the lowest bit set.

## Why?

This allows us to trace through code that calls an ephemeral function, i.e., a function that may not be alive when we are constructing the executor, e.g. a generator expression or certain nested functions. We will lose globals removal inside such functions, but we can still do other peephole operations (and even possibly [call inlining](https://github.com/python/cpython/pull/116290), if we decide to do it), which only need the code object. As before, if we cannot retrieve the code object from the cache, we stop projecting.

For more details see the issue.

* Issue: gh-117045

### CODE REVIEW

**Comment by gvanrossum** [comment]
Performance is neutral:
https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240319-3.13.0a5%2B-0fd96be-JIT/README.md

**Comment by gvanrossum** [comment]
@markshannon, @brandtbucher, please review, and please let me know whether you want any of the [API design changes](https://github.com/python/cpython/issues/117045#issuecomment-2008188497) from the issue implemented.

File: Include/internal/pycore_frame.h [hunk 1]
```diff
@@ -55,7 +55,7 @@ enum _frameowner {
 };
 
 typedef struct _PyInterpreterFrame {
-    PyObject *f_executable; /* Strong reference */
+    PyObject *f_executable; /* Strong reference (code object) */
```
Comment (markshannon) [inline Include/internal/pycore_frame.h]: Not necessarily. It is `None` for entry frames.

File: Objects/funcobject.c [hunk 2]
```diff
@@ -218,92 +218,131 @@ PyFunction_NewWithQualName(PyObject *code, PyObject *globals, PyObject *qualname
...
+        assert(slot->func->func_code == slot->code);
+        return slot->func;
     }
     return NULL;
 }
 
 uint32_t
 _PyFunction_GetVersionForCurrentState(PyFunctionObject *func)
 {
-    if (func->func_version != 0) {
```
Comment (markshannon) [inline Objects/funcobject.c:294]: I think this is fine.
Are there any changes made to functions in decorators like `functool.wraps` that would mess this up?

File: Python/bytecodes.c [hunk 3]
```diff
@@ -3218,7 +3218,6 @@ dummy_func(
             assert(func->func_defaults);
             assert(PyTuple_CheckExact(func->func_defaults));
             int defcount = (int)PyTuple_GET_SIZE(func->func_defaults);
-            assert(defcount <= code->co_argcount);
```
Comment (markshannon) [inline Python/bytecodes.c]: Please leave this. The assertion is correct, it is the optimization that is (currently) faulty.

File: Python/optimizer_analysis.c [hunk 4]
```diff
@@ -225,7 +225,12 @@ remove_globals(_PyInterpreterFrame *frame, _PyUOpInstruction *buffer,
                 builtins_watched <<= 1;
                 globals_watched <<= 1;
                 function_checked <<= 1;
-                PyFunctionObject *func = (PyFunctionObject *)buffer[pc].operand;
+                uintptr_t operand = buffer[pc].operand;
```
Comment (markshannon) [inline Python/optimizer_analysis.c]:
```suggestion
                uint64_t operand = buffer[pc].operand;
```

File: Python/optimizer_analysis.c [hunk 5]
```diff
@@ -247,7 +252,15 @@ remove_globals(_PyInterpreterFrame *frame, _PyUOpInstruction *buffer,
                 builtins_watched >>= 1;
                 globals_watched >>= 1;
                 function_checked >>= 1;
-                PyFunctionObject *func = (PyFunctionObject *)buffer[pc].operand;
+                uintptr_t operand = buffer[pc].operand;
```
Comment (markshannon) [inline Python/optimizer_analysis.c]:
```suggestion
                uint64_t operand = buffer[pc].operand;
```

File: Python/optimizer_bytecodes.c [hunk 6]
```diff
@@ -543,12 +543,23 @@ dummy_func(void) {
...
+        PyCodeObject *co = NULL;
+        assert((this_instr + 2)->opcode == _PUSH_FRAME);
+        uintptr_t push_operand = (this_instr + 2)->operand;
+        if (push_operand & 1) {
+            co = (PyCodeObject *)(push_operand & ~1);
+            DPRINTF(3, "code=%p ", co);
+            assert(PyCode_Check(co));
+        }
+        else {
+            PyFunctionObject *func = (PyFunctionObject *)(this_instr + 2)->operand;
```
Comment (markshannon) [inline Python/optimizer_bytecodes.c]:
```suggestion
            PyFunctionObject *func = (PyFunctionObject *)push_operand;
```

**Review by markshannon — COMMENTED** [review]
A few comments, but looks good in general.

File: Objects/funcobject.c (hunk 2 above)
Comment (gvanrossum) [reply Objects/funcobject.c:294]: No, that leaves all the critical attributes alone.

**Comment by gvanrossum** [comment]
We are still inconsistent in how we cast to or from `operand` -- some places still use `uintptr_t`. I guess we can clean that up as we touch that code.

**Comment by gvanrossum** [comment]
Argh, CI is failing because of problems with launchpad. I'll rerun them later.

**Comment by gvanrossum** [comment]
Closing and reopening to restart the tests. (I can't seem to rerun just the failed tests in GH somehow.)

**Pull request merged**